File: clients/ios/RemoteEverything/Core/Update/UpdateChecker.swift
import Foundation

/// What this build says about itself. The version and the build number come from
/// the bundle's Info.plist, which clients/release.json is the source of; the
/// protocol version is the one this source speaks.
enum AppVersion {
    static let protocolVersion = 1

    static var versionName: String {
        Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "0.0.0"
    }

    static var buildNumber: Int {
        guard let raw = Bundle.main.object(forInfoDictionaryKey: "CFBundleVersion") as? String else { return 0 }
        return Int(raw) ?? 0
    }
}

/// What checking for updates found.
enum UpdateResult: Equatable {
    case upToDate
    case available(versionName: String, buildNumber: Int)
    /// The protocol itself changed: upgrading is not enough, pairing again is
    /// part of it. There is no negotiation and no compatibility path.
    case protocolChanged(protocolVersion: Int)
    case unreachable

    var isProtocolChange: Bool {
        if case .protocolChanged = self { return true }
        return false
    }
}

/// Checking for a newer release. Deliberately decoupled from any gateway: a
/// device that has never paired can still check, and a gateway being down must
/// not look like a version problem.
final class UpdateChecker {

    /// The published release manifest — a fixed address, not a gateway path.
    static let manifestURLString = "https://raw.githubusercontent.com/hxaxd/remote-everything/main/clients/release.json"

    private let session: URLSession
    private let manifestURLString: String

    init(session: URLSession? = nil, manifestURLString: String = UpdateChecker.manifestURLString) {
        if let session {
            self.session = session
        } else {
            let configuration = URLSessionConfiguration.ephemeral
            // A manifest is not a probe and not a gateway: it gets the same time
            // as any other request this client makes, `Cadence`'s, and no more.
            configuration.timeoutIntervalForRequest = Cadence.requestTimeout
            configuration.timeoutIntervalForResource = Cadence.requestTimeout * 2
            configuration.waitsForConnectivity = false
            self.session = URLSession(configuration: configuration)
        }
        self.manifestURLString = manifestURLString
    }

    func check(currentBuildNumber: Int, currentProtocolVersion: Int) async -> UpdateResult {
        guard let url = URL(string: manifestURLString) else { return .unreachable }
        var request = URLRequest(url: url)
        request.timeoutInterval = Cadence.requestTimeout
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        guard let (data, response) = try? await session.data(for: request) else { return .unreachable }
        guard let http = response as? HTTPURLResponse, http.statusCode == 200 else { return .unreachable }
        guard let manifest = try? JSONDecoder().decode(ReleaseManifest.self, from: data) else { return .unreachable }

        if manifest.protocolVersion != currentProtocolVersion {
            return .protocolChanged(protocolVersion: manifest.protocolVersion)
        }
        if manifest.buildNumber > currentBuildNumber {
            return .available(versionName: manifest.versionName, buildNumber: manifest.buildNumber)
        }
        return .upToDate
    }
}

File: clients/android/app/src/main/java/com/remoteeverything/core/update/UpdateChecker.kt
package com.remoteeverything.core.update

import com.remoteeverything.core.json.Strict
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.OkHttpClient
import okhttp3.Request
import java.util.concurrent.TimeUnit

/**
 * Update check against the published release manifest. Deliberately decoupled
 * from any gateway: an unpaired device can check too. The manifest is the
 * single source of truth all three platforms are versioned from.
 */
class UpdateChecker(
    private val manifestUrl: String = DEFAULT_MANIFEST_URL,
    private val client: OkHttpClient = defaultClient(),
) {

    @Serializable
    data class ReleaseManifest(
        val schema: Int,
        val versionName: String,
        val buildNumber: Int,
        val protocolVersion: Int,
        val minimumPlatforms: MinimumPlatforms,
    )

    @Serializable
    data class MinimumPlatforms(
        val androidSdk: Int,
        val ios: String,
        val harmonyApi: String,
    )

    sealed class Result {
        data object UpToDate : Result()
        data class Available(val versionName: String, val buildNumber: Int) : Result()
        data class ProtocolChanged(val protocolVersion: Int) : Result()
        data object Unreachable : Result()
    }

    private val json = Json { ignoreUnknownKeys = false }

    suspend fun check(currentBuildNumber: Int, currentProtocolVersion: Int): Result = withContext(Dispatchers.IO) {
        val request = Request.Builder().url(manifestUrl).build()
        val body = try {
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) return@withContext Result.Unreachable
                response.body.string()
            }
        } catch (e: Exception) {
            return@withContext Result.Unreachable
        }
        val manifest = try {
            json.decodeFromString<ReleaseManifest>(body)
        } catch (e: Exception) {
            return@withContext Result.Unreachable
        }
        val platforms = setOf(
            "androidSdk", "ios", "harmonyApi",
        )
        try {
            Strict.validateRelease(
                manifest.schema,
                manifest.versionName,
                manifest.buildNumber,
                manifest.protocolVersion,
                platforms,
            )
        } catch (e: Exception) {
            return@withContext Result.Unreachable
        }
        when {
            manifest.protocolVersion != currentProtocolVersion -> Result.ProtocolChanged(manifest.protocolVersion)
            manifest.buildNumber > currentBuildNumber -> Result.Available(manifest.versionName, manifest.buildNumber)
            else -> Result.UpToDate
        }
    }

    companion object {
        const val DEFAULT_MANIFEST_URL =
            "https://raw.githubusercontent.com/hxaxd/remote-everything/main/clients/release.json"

        private fun defaultClient(): OkHttpClient = OkHttpClient.Builder()
            .connectTimeout(10, TimeUnit.SECONDS)
            .readTimeout(10, TimeUnit.SECONDS)
            .build()
    }
}
